MonitorControl/UI/SliderHandler.swift

```diff
@@ -45,13 +46,130 @@ class SliderHandler {
         self.display.saveValue(brightnessDDCValue, for: self.cmd)
+      } else if self.cmd == Utils.Command.audioSpeakerVolume {
+        if !self.display.enableMuteUnmute || value != 0 {
+          _ = self.display.writeDDCValues(command: self.cmd, value: UInt16(value))
+        }
+        self.display.saveValue(value, for: self.cmd)
       } else {
         _ = self.display.writeDDCValues(command: self.cmd, value: UInt16(value))
         self.display.saveValue(value, for: self.cmd)
       }
-    } else if self.cmd == DDC.Command.brightness {
+    } else if self.cmd == Utils.Command.brightness {
       _ = self.display.setSwBrightness(value: UInt8(value))
       self.display.saveValue(value, for: self.cmd)
     }
   }
+
+  static func addSliderMenuItem(toMenu menu: NSMenu, forDisplay display: ExternalDisplay, command: Utils.Command, title: String, numOfTickMarks: Int = 0) -> SliderHandler {
+    let item = NSMenuItem()
+
+    let handler = SliderHandler(display: display, command: command)
+
+    let slider = NSSlider(value: 0, minValue: 0, maxValue: 100, target: handler, action: #selector(SliderHandler.valueChanged))
+    slider.isEnabled = false
+    handler.slider = slider
+
+    if #available(macOS 11.0, *) {
+      slider.frame.size.width = 160
+      slider.frame.origin = NSPoint(x: 35, y: 5)
+      let view = NSView(frame: NSRect(x: 0, y: 0, width: slider.frame.width + 47, height: slider.frame.height + 14))
+      view.frame.origin = NSPoint(x: 12, y: 0)
+      var iconName: String = "circle.dashed"
+      switch command {
+      case .audioSpeakerVolume: iconName = "speaker.wave.2"
+      case .brightness: iconName = "sun.max"
+      case .contrast: iconName = "circle.lefthalf.fill"
+      default: break
+      }
+      let icon = NSImageView(image: NSImage(systemSymbolName: iconName, accessibilityDescription: title)!)
+      icon.frame = view.frame
+      icon.wantsLayer = true
+      icon.alphaValue = 0.7
+      icon.imageAlignment = NSImageAlignment.alignLeft
+      view.addSubview(icon)
+      view.addSubview(slider)
+      item.view = view
+      menu.insertItem(item, at: 0)
+    } else {
+      slider.frame.size.width = 180
+      slider.frame.origin = NSPoint(x: 15, y: 5)
+      let view = NSView(frame: NSRect(x: 0, y: 0, width: slider.frame.width + 30, height: slider.frame.height + 10))
+      let sliderHeaderItem = NSMenuItem()
+      let attrs: [NSAttributedString.Key: Any] = [.foregroundColor: NSColor.systemGray, .font: NSFont.systemFont(ofSize: 12)]
+      sliderHeaderItem.attributedTitle = NSAttributedString(string: title, attributes: attrs)
+      view.addSubview(slider)
+      item.view = view
+      menu.insertItem(item, at: 0)
+      menu.insertItem(sliderHeaderItem, at: 0)
+    }
+
+    var values: (UInt16, UInt16)?
+    let delay = display.needsLongerDelay ? UInt64(40 * kMillisecondScale) : nil
+
+    let tries = UInt(display.getPollingCount())
+    os_log("Polling %{public}@ times", type: .info, String(tries))
```

Comment:
These lines should be moved inside the if condition where the safemode check is imo, I still noticed the message "Polling x times" in the logs when launching through safe mode. Technically this doesn't matter but it won't hurt to move them to the right spot either imo (especially for debugging).

Reply:
You are right, I did change this one (commited, pushed).